crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+                };
+                info!("[JWK] qc update obtained, update={:?}", update);
+                Ok(())
+            },
+            _ => Err(anyhow!(
+                "qc update not expected for issuer {:?} in state {}",
+                update.update.issuer,
+                state.consensus_state.name()
+            )),
+        }
+    }
+}
+
+/// An instance of this resource is created when `JWKManager` starts the QC update building process for an issuer.
+/// Then `JWKManager` needs to hold it. Once this resource is dropped, the corresponding QC update process will be cancelled.
+#[derive(Clone, Debug)]
+pub struct QuorumCertProcessGuard {
+    handle: Option<AbortHandle>,
+}
+
+impl QuorumCertProcessGuard {
+    pub fn new(handle: AbortHandle) -> Self {
+        Self {
+            handle: Some(handle),
+        }
+    }
+
+    #[cfg(test)]
+    pub fn dummy() -> Self {
+        Self { handle: None }
+    }
+}
+
+impl Drop for QuorumCertProcessGuard {
+    fn drop(&mut self) {
+        let QuorumCertProcessGuard { handle } = self;
+        if let Some(handle) = handle {
+            handle.abort();
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum ConsensusState {
+    NotStarted,
+    InProgress {
+        my_proposal: ObservedUpdate,
+        abort_handle_wrapper: QuorumCertProcessGuard,
+    },
+    Finished {
+        vtxn_guard: TxnGuard,
+        my_proposal: ObservedUpdate,
+        quorum_certified: QuorumCertifiedUpdate,
+    },
+}
+
+impl PartialEq for ConsensusState {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (ConsensusState::NotStarted, ConsensusState::NotStarted) => true,
+            (
+                ConsensusState::InProgress {
+                    my_proposal: update_0,
+                    ..
+                },
+                ConsensusState::InProgress {
+                    my_proposal: update_1,
+                    ..
+                },
+            ) if update_0 == update_1 => true,
+            (
+                ConsensusState::Finished {
+                    my_proposal: update_0,
+                    ..
+                },
+                ConsensusState::Finished {
+                    my_proposal: update_1,
+                    ..
+                },
+            ) if update_0 == update_1 => true,
+            _ => false,
+        }
+    }
+}
+
+impl Eq for ConsensusState {}
+
+impl ConsensusState {
+    pub fn name(&self) -> &str {
+        match self {
+            ConsensusState::NotStarted => "NotStarted",
+            ConsensusState::InProgress { .. } => "InProgress",
+            ConsensusState::Finished { .. } => "Finished",
+        }
+    }
+
+    pub fn my_proposal_cloned(&self) -> ObservedUpdate {
+        match self {
+            ConsensusState::InProgress { my_proposal, .. }
+            | ConsensusState::Finished { my_proposal, .. } => my_proposal.clone(),
+            _ => panic!("my_proposal unavailable"),
```

Comment:
Is this only used in the tests? Can we mark it test-only, or return option and panic in the test?

Reply:
fixed